tools/common_lib/src/gemm.h

```diff
@@ -713,13 +969,197 @@ class GemmBaseDispatcher : public NodeDispatcher
             barriers.push_back(CD3DX12_RESOURCE_BARRIER::Transition(input_buffer_b_.Get(),
                 D3D12_RESOURCE_STATE_COPY_DEST, D3D12_RESOURCE_STATE_UNORDERED_ACCESS));
         }
-        if (input_buffer_c_)
+        if (input_buffer_c_ && tensor_input_c_bytes_width > 0)
         {
             barriers.push_back(CD3DX12_RESOURCE_BARRIER::Transition(input_buffer_c_.Get(),
                 D3D12_RESOURCE_STATE_COPY_DEST, D3D12_RESOURCE_STATE_UNORDERED_ACCESS));
         }
+        if (input_buffer_scale_)
+        {
+            barriers.push_back(CD3DX12_RESOURCE_BARRIER::Transition(input_buffer_scale_.Get(),
+                D3D12_RESOURCE_STATE_COPY_DEST, D3D12_RESOURCE_STATE_UNORDERED_ACCESS));
+        }
+        if (input_buffer_zeropoint_)
+        {
+            barriers.push_back(CD3DX12_RESOURCE_BARRIER::Transition(input_buffer_zeropoint_.Get(),
+                D3D12_RESOURCE_STATE_COPY_DEST, D3D12_RESOURCE_STATE_UNORDERED_ACCESS));
+        }
         cmd_list->ResourceBarrier(static_cast<std::uint32_t>(barriers.size()), barriers.data());
     }
+    
+
+    GemmBaseDispatcher(create_params_t&& params, ID3D12Device* d3d12_device, IDMLDevice* dml_device, IDMLCommandRecorder* dml_cmd_recorder, ID3D12GraphicsCommandList* cmd_list)
+        : params_(std::move(params))
+        , dml_cmd_recorder_(dml_cmd_recorder)
+        , dml_device_(dml_device)
+        , d3d12_device_(d3d12_device)
+
+    {
+        if (params_.type == GemmType::GemmType_Quant)
```

Comment:
This is really convoluted design. We have condition here which calls some ctor, which inside calls some DML base class which again is checking if it's quantized gemm. 
Please keep it simple and just create fully separated dispatcher for qGEMM or keep extending ordinary "AB" gemm.

Reply:
Have restructured the code and made Quant Gemm to be a separate class.